x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Categorize.java

```diff
@@ -156,4 +176,175 @@ public Expression field() {
     public String toString() {
         return "Categorize{field=" + field + "}";
     }
+
+    public GroupingKey.Supplier groupingKey(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        return mode -> new GroupingKeyFactory(source(), toEvaluator.apply(field), mode);
+    }
+
+    record GroupingKeyFactory(Source source, ExpressionEvaluator.Factory field, AggregatorMode mode) implements GroupingKey.Factory {
+        @Override
+        public GroupingKey apply(DriverContext context, int resultOffset) {
+            ExpressionEvaluator field = this.field.get(context);
```

Comment:
Why do we need an expression evaluator here? So we can compute expressions like `STATS ... BY CATEGORIZE(SUBSTRING(field, 10))`?

If `CATEGORIZE` is only run as part of a `STATS`, this may not be the best way to generalize it to nested expressions; for aggs, our current way is to extract nested expressions into an EVAL - like `EVAL $$temp = SUBSTRING(field, 10) | STATS ... BY CATEGORIZE($$temp)`.

**Related and important:** Currently our optimizer expects the `BY ...` groups to contain only field and reference attributes after the initial substitution step; I understand that `CATEGORIZE` breaks this assumption, so we'll have to go and update some optimizer rules to expect references, fields _and_ `CATEGORIZE` in the groupings.

Reply:
Bleh, never mind. I realized we require running the initial categorization as an Eval and the second one as part of the final Aggregate, like so

```
Coordinator:
AggregateExec[[categorize(x){r}#57],[COUNT([2a][KEYWORD]) AS count(*), categorize(x){r}#57],FINAL,[categorize(x){r}#57, count{r
}#62, seen{r}#63],58]
\_ExchangeSourceExec[[categorize(x){r}#57, count{r}#62, seen{r}#63],true]

Local:
ExchangeSinkExec[[categorize(x){r}#72, count{r}#77, seen{r}#78],true]
\_AggregateExec[[categorize(x){r}#72],[COUNT([2a][KEYWORD]) AS count(*), categorize(x){r}#72],INITIAL,[categorize(x){r}#72, count
{r}#81, seen{r}#82],58]
  \_EvalExec[[CATEGORIZE(x{f}#75) AS categorize(x)]]
    \_FieldExtractExec[x{f}#75]<[]>
      \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#83], limit[], sort[] estimatedRowSize[104]
```